### PR TITLE
Metadata v8

### PR DESCRIPTION
**Pull request by regit**
Update of #2950 addressing the comments.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- introduce a hash table to store the key and value
- rename option
- conditional parsing

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/343
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/127

### CODE REVIEW

File: src/detect-metadata.c [hunk 1]
```diff
@@ -28,23 +28,271 @@
...
+
+        dkv = SCMalloc(sizeof(DetectMetadata));
+        if (dkv) {
+            dkv->key = key;
+            if (!dkv->key) {
+                SCFree(dkv);
+            } else {
+                dkv->value = value;
+                if (!dkv->value) {
+                    SCFree((void *)dkv->key);
```
Comment (victorjulien) [inline src/detect-metadata.c]: is the key not owned by the hash table?

File: src/detect-metadata.c [hunk 2]
```diff
@@ -28,23 +28,271 @@
...
+            goto error;
+        }
+
+        kv = strtok_r(NULL, ",", &saveptr);
+    } while (kv);
+
+    return 0;
+
+error:
+    if (key)
```
Comment (victorjulien) [inline src/detect-metadata.c]: when does this happen? Are key and value not owned by the hash table?

File: src/detect-metadata.c [hunk 3]
```diff
@@ -28,23 +28,271 @@
...
+    FAIL_IF(strcmp("gros_minet", sig->metadata->value));
+    FAIL_IF_NULL(sig->metadata->next); 
+    dm = sig->metadata->next;
+    FAIL_IF(strcmp("titi", dm->key));
+    dm = dm->next;
+    FAIL_IF_NULL(dm);
+    FAIL_IF(strcmp("toto", dm->key));
+
+    DetectEngineCtxFree(de_ctx);
+    PASS;
```
Comment (victorjulien) [inline src/detect-metadata.c]: should DetectEngineSetParseMetadata be reset here? Ideally a test doesn't modify global state

Comment (victorjulien) [reply src/detect-metadata.c]: not permanently I mean

File: src/detect-metadata.h [hunk 4]
```diff
@@ -24,8 +24,22 @@
 #ifndef __DETECT_METADATA_H__
 #define __DETECT_METADATA_H__
 
+/**
+ * \brief Signature metadata list.
+ */
+typedef struct DetectMetadata_ {
+    /* pointer to key */
```
Comment (victorjulien) [inline src/detect-metadata.h]: can you add explanation on who owns these ptrs (hash table as far as I understand)

**Comment by regit** [comment]
Oups, forgot the error handling. Gonna resubmit.

**Pull request closed without merging**